scheduler: align with ctrl-managers and apiservers, add https+auth in options

sttts opened the pull request:
This switches the bootstrapping code to use the Options+Config pattern which
allows to compose flag sets and flag logic from reusable components.

This way https+authz+authn support is added to the Options struct (still disabled
for the actual binary, but available for downstreams).

TODOs:
- [x] fix kubeadm

sttts committed:
scheduler: align plumbing with controller-manager and apiservers

sttts committed:
controller-manager: make InsecureServingOptions/Config re-usable

sttts committed:
scheduler: add https+authn+authz to options, set to nil for now

sttts committed:
controller-manager: unify address flag description to listen on all interfaces

sttts committed:
Update bazel

sttts commented:
/cc @luxas 

For kubeadm commit.

sttts commented:
/assign @hzxuzhonghu

hzxuzhonghu commented on cmd/controller-manager/app/insecure_serving.go:
```diff
@@ -30,6 +30,8 @@ import (
 	// Listener is the secure server network listener.
 	Listener net.Listener
+	// optional server name for log messages
+	Name string
```
Do we really need this？ kube-controller and cloud-controller start up as two separate processes.

hzxuzhonghu replied:
ignore this, used for distinct  metrics and healthz

hzxuzhonghu commented on cmd/kube-scheduler/app/config/config.go:
```diff
@@ -0,0 +1,74 @@
+		c.InsecureServing.Name = "healthz"
+	}
+	if c.InsecureMetricsServing != nil {
+		c.InsecureServing.Name = "metrics"
```
InsecureMetricsServing ?

sttts replied:
yes, on it. Revamping the whole logic right now. With tests!

sttts replied:
done

hzxuzhonghu replied:
Do not see changes.

sttts replied:
misread your comment. Fixed now.

hzxuzhonghu commented on cmd/kube-scheduler/app/options/configfile.go:
```diff
@@ -0,0 +1,53 @@
+	componentconfigv1alpha1 "k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1"
+)
+
+func loadConfigFromFile(file string) (*componentconfig.KubeSchedulerConfiguration, error) {
```
Should this fit for kube-controller-manager?

sttts replied:
no, not yet. But eventually we need the same for the controller manager.

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
2018

sttts replied:
fixed

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,234 @@
+limitations under the License.
+*/
+
+// Package app implements a Server object for running the scheduler.
```
nit: options

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,234 @@
+	// The default values. These are overridden if ConfigFile is set or by values in InsecureServing.
+	ComponentConfig componentconfig.KubeSchedulerConfiguration
+
+	InsecureServing        *controlleroptions.InsecureServingOptions // these override the ConfigFile values if set
```
`InsecureServingOptions`  should be put into a more common place.

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,234 @@
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	controlleroptions "k8s.io/kubernetes/cmd/controller-manager/app/options"
```
Weird

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,234 @@
+	o := &Options{
+		ComponentConfig: *cfg,
+		InsecureServing: &controlleroptions.InsecureServingOptions{
+			BindPort:    -1, // invalid port, updated from ComponentConfig or set to zero in ApplyTo
```
why not leave it to default 0.

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,234 @@
+
+			// sync back ComponentConfig.HealthzBindAddress to o.InsecureServing
+			if len(o.ComponentConfig.HealthzBindAddress) == 0 {
+				o.InsecureServing.BindPort = 0
```
and remove this branch

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,234 @@
+				return fmt.Errorf("invalid non-numerical metrics port %q", port)
+			}
+			metricsServing.BindAddress = net.ParseIP(host)
+			metricsServing.BindPort = portInt
```
if metricsServing and healthServing have same port but different BindAddress , will cause a conflict?

hzxuzhonghu commented on cmd/kube-scheduler/app/options/plumbing.go:
```diff
@@ -0,0 +1,155 @@
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	apimachineryerrors "k8s.io/apimachinery/pkg/util/errors"
```
alias: utilerrors

sttts replied:
done

hzxuzhonghu commented on cmd/kube-scheduler/app/options/plumbing.go:
```diff
@@ -0,0 +1,155 @@
+	"github.com/golang/glog"
+)
+
+// Config return a cloud controller manager config objective
```
scheduler

sttts replied:
done

hzxuzhonghu commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,240 @@
+
+	isPort := is.BindPort
+	isAddress := is.BindAddress
+	if is.Listener != nil {
```
Not necessary branch the Listener is definitely nil .  Not set before.

sttts replied:
Tests can set it.

hzxuzhonghu commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,240 @@
+	return nil
+}
+
+func syncBindAddressAndInsecureServingOptions(addr *string, is *controlleroptions.InsecureServingOptions) error {
```
Am not sure whether this really useful.

hzxuzhonghu replied:
Doe not `applyDeprecatedHealthzAddressToConfig` and `applyDeprecatedHealthzPortToConfig` already set the addr and port ?

sttts replied:
Because we have bad situation that we have a non-modular, non-reusable scheduler component config, and we have componentized option structs. If we split the former in a similar way we can add native ComponentConfigInsecureServing support to the options. Until then we need this sync logic because the flags, Options and component config are sources for the configuration, with certain rules. This leads to quite some legacy complexity.

sttts replied:
Documented this function and added tests.

mikedanese commented:
/retest

deads2k commented on cmd/kube-scheduler/app/config/config.go:
```diff
@@ -0,0 +1,78 @@
+// Config has all the context to run a Scheduler
+type Config struct {
+	// config is the scheduler server's configuration object.
+	ComponentConfig componentconfig.KubeSchedulerConfiguration
```
@aveshagarwal @kubernetes/sig-scheduling-feature-requests we need to figure out what is local to the machine the scheduler is running on and what is shared configuration and placed in etcd.  It's not obvious to me where that line is drawn and some things seem unusual.  For instance, there is a file reference for a kubeconfigfile that refers to something on disk.  1. the kubeconfig is needed in order to contact the apiserver to find the configmap config, and 2. I don't think I would expect the disk layout across all schedulers to be consistent.

deads2k replied:
Continuing that thought, I wouldn't expect the bind addresses to be shared either.

I think I would that I want configuration options controlling behavior to be shared amongst all scheduler instances and I want configuration options controlling exposure and connections to be specific to a particular scheduler instance.

This pull has the misfortune of exposing these problems.

aveshagarwal replied:
@deads2k so far there are 2 things the scheduler is referring on disk: 1)  https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/componentconfig/types.go#L47 which is one of options among on-disk file, configmap and algorithmprovider 2) path to kubeconfig:  https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/componentconfig/types.go#L26 

> I don't think I would expect the disk layout across all schedulers to be consistent.

makes sense.

> I think I would that I want configuration options controlling behavior to be shared amongst all scheduler instances and I want configuration options controlling exposure and connections to be specific to a particular scheduler instance.

So that means these options kubeconfigfile path, bindaddresses (health, metrics), QPS, Burst  per instance, and i think all other options can be shared.

deads2k commented:
It's not clear to me where and how we'll draw lines between shared config (assuming config in etcd is intended to be shared) and instance specific config (which should still be held in a config struct, but which needs to be local).

@ncdc you were involved in this config struct.  Is there a plan?

ncdc commented:
@deads2k I only really did kube-proxy, and gave the scheduler shift a cursory glance (not enough time).

However, I did have a brief gh chat with @mtaufen about this, and I think he was leaning towards using a single ConfigMap for all common attributes, and retaining flags for instance specific ones.

sttts commented:
> a single ConfigMap

A single config file is a good replacement for flags. But as soon as we move to ConfigMaps we have to merge the file and the ConfigMap, or allow flags to override the ConfigMap. I fear a lot of non-trivial merge logic there in the former case of file+map. Using flags pointing to the internal type fields of the component config type is much easier.

I agree with @deads2k that having clearly local fields like filepaths, bind ports and addresses in the ConfigMap is strange.

ncdc commented:
> But as soon as we move to ConfigMaps we have to merge the file and the ConfigMap

These are one and the same. The file comes from the ConfigMap (volume mount). There is nothing to merge here.

> or allow flags to override the ConfigMap

NO! The flags are being deprecated, replaced by the config file (which, to reiterate, comes from the ConfigMap).

> I agree with @deads2k that having clearly local fields like filepaths, bind ports and addresses in the ConfigMap is strange.

TBH, I only replied to David's question about common vs. per-instance settings. I didn't look at the specifics like file paths.

sttts commented:
> These are one and the same.
> NO! The flags are being deprecated

Which then forces us to have all flags in the config, independently from being local or not. Feels odd.

ncdc commented:
> Which then forces us to have all flags in the config, independently from being local or not. Feels odd.

TBH local vs not wasn't something I considered when doing the kube-proxy config. I also didn't consider per-instance settings. I would say the latter is probably a poor oversight, but I don't necessarily think that having a yaml file in a ConfigMap point to a file on the local filesystem is a terrible thing, if you install all your instances consistently.

There are probably some options in this area:

1) Don't allow any flags other than `--config-file`. This means you need a separate config file (and ConfigMap) per instance. Not desirable, since presumably 80-90% or more of the configuration is common across all instances.

2) Have `--config-file` and `--instance-config-file`. Put the common stuff in the former, and the instance specific stuff in the latter. Probably more confusing than having just a single config file. Maybe better names would help.

3) Have `--config-file` and per-instance flags.

We also have some options for references to local resources. We could say they're never in the common config file. We could say they have to be flags. We could say they go in the per-instance config file.

We probably need to have this discussion somewhere other than this issue 😄

mtaufen commented:
> I don't necessarily think that having a yaml file in a ConfigMap point to a file on the local filesystem is a terrible thing

This should absolutely be possible. Additionally, relative paths should be resolved from the config file's location, so that you can do this:

```
kind: ConfigMap
data:
  some-file: "stuff"
  component: {\"pathToStuff\": \"some-file\"}
```

(1) could work if you can auto-discover all instance-specific configuration at runtime, in which case you could get rid of the instance-specific flags. If you can't auto-discover, (1) is a non-starter.

(2) may be where we eventually land, given we could provide the same level of API guarantee across all config sources.

(3) is where we are today with the Kubelet, which is functional but unsatisfying from a unified-machinery standpoint.

mtaufen commented:
Also just want to add that kube-proxy and kubelet standardized on `--config` for the flag that points to the config file, so we should try to stick to that naming as much as possible.

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,173 @@
+	if err := o.SecureServing.ApplyTo(&c.SecureServing); err != nil {
+		return err
+	}
+	if err := o.CombinedInsecureServing.ApplyTo(c, &c.ComponentConfig); err != nil {
```
Is this duplicated with line 137 and 145?

sttts replied:
rebase error. Fixing.

sttts replied:
Fixed

hzxuzhonghu replied:
clear and readable now

hzxuzhonghu commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,171 @@
+		if cfg, err := loadConfigFromFile(o.ConfigFile); err != nil {
+			return err
+		} else {
+			c.ComponentConfig = *cfg
```
bug here cfg undefined

sttts replied:
nope, fixed in this push. Was a bug before.

hzxuzhonghu replied:
this caused golint error

hzxuzhonghu commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,167 @@
+		return nil
+	}
+
+	if err := updateAddressFromInsecureServingOptions(&componentConfig.HealthzBindAddress, o.Healthz); err != nil {
```
redundant for `ApplyToFromLoadedConfig`

sttts replied:
Why redundant?

hzxuzhonghu replied:
`HealthzBindAddress` is already loaded from config file, no need to write back again.

hzxuzhonghu replied:
ApplyToFromLoadedConfig already use it.
`if err := updateInsecureServingOptionsFromAddress(o.Healthz, componentConfig.HealthzBindAddress); err != nil {`

sttts replied:
not redundant because `updateAddressFromInsecureServingOptions` switches off the insecure serving if the respective `InsecureServingOptions` is nil (we have a test case for that which breaks if you remove those calls).

hzxuzhonghu commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,167 @@
+	if err := updateAddressFromInsecureServingOptions(&componentConfig.HealthzBindAddress, o.Healthz); err != nil {
+		return err
+	}
+	if err := updateAddressFromInsecureServingOptions(&componentConfig.MetricsBindAddress, o.Metrics); err != nil {
```
same as above

sttts replied:
same as above

hzxuzhonghu commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,167 @@
+	if err := o.Healthz.ApplyTo(&c.InsecureServing); err != nil {
+		return err
+	}
+	if o.Metrics != nil && (c.ComponentConfig.MetricsBindAddress != c.ComponentConfig.HealthzBindAddress || o.Healthz == nil) {
```
I still have a doubt if MetricsBindAddress and HealthzBindAddress are loaded from file and they have different address but same port. Bootstrap will fail.

hzxuzhonghu replied:
e.g. MetricsBindAddress = "0.0.0.0:8080"
HealthzBindAddress = "127.0.0.1:8080"

sttts replied:
no. this works fine.

There might be different ways to write addresses for ipv6. But this is out of our control and an error in the logs is fine.

I am only replicating the old logic here. I don't like the way it is done either as there is too much magic involved.

hzxuzhonghu replied:
yes, it is old logic.

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60372)
<!-- Reviewable:end -->

sttts commented:
@luxas approved?

sttts commented:
@aveshagarwal can you take a look? This is preparation for tls+authn+authc support, needed for 3.11.
@k82cn is this approved?

aveshagarwal commented:
@sttts will review it soon.

k82cn commented:
> @k82cn is this approved?

Sorry for the late, let me check it this week :)

k82cn commented:
/assign

sttts commented:
> let me check it this week

thanks :)

aveshagarwal commented on cmd/controller-manager/app/options/insecure_serving.go:
```diff
@@ -92,8 +85,12 @@ func (s *InsecureServingOptions) ApplyTo(c **genericcontrollermanager.InsecureSe
 
 	if s.Listener == nil {
 		var err error
+		listen := options.CreateListener
```
could you move it to else part in the if below?

sttts replied:
Would leave this short here, avoiding a declaration.

aveshagarwal commented on cmd/kube-scheduler/app/options/configfile.go:
```diff
@@ -0,0 +1,53 @@
+	scheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(scheme)
+	componentconfig.AddToScheme(scheme)
+	componentconfigv1alpha1.AddToScheme(scheme)
```
would it be better to move the above to some sort of init function?

aveshagarwal commented on cmd/kube-scheduler/app/options/deprecated.go:
```diff
@@ -0,0 +1,116 @@
+	fs.StringVar(&o.AlgorithmProvider, "algorithm-provider", o.AlgorithmProvider, "DEPRECATED: the scheduling algorithm provider to use, one of: "+factory.ListAlgorithmProviders())
+	fs.StringVar(&o.PolicyConfigFile, "policy-config-file", o.PolicyConfigFile, "DEPRECATED: file with scheduler policy configuration. This file is used if policy ConfigMap is not provided or --use-legacy-policy-config==true")
+	usage := fmt.Sprintf("DEPRECATED: name of the ConfigMap object that contains scheduler's policy configuration. It must exist in the system namespace before scheduler initialization if --use-legacy-policy-config==false. The config must be provided as the value of an element in 'Data' map with the key='%v'", componentconfig.SchedulerPolicyConfigMapKey)
+	fs.StringVar(&o.PolicyConfigMapName, "policy-configmap", o.PolicyConfigMapName, usage)
```
In the flood of deprecated flags, it really took me some time to find this loner.

aveshagarwal replied:
ok it seems that it has usage above that shows DEPRECATED. nvm.

sttts replied:
Just copied from the old code.

aveshagarwal commented on cmd/kube-scheduler/app/config/config.go:
```diff
@@ -0,0 +1,78 @@
+	InformerFactory                informers.SharedInformerFactory
+	PodInformer                    coreinformers.PodInformer
+	AlgorithmSource                componentconfig.SchedulerAlgorithmSource
+	HardPodAffinitySymmetricWeight int32
```
Wondering why SchedulerName, AlgorithmSource and HardPodAffinitySymmetricWeight are in this list again when they are already part of ComponentConfig above?

aveshagarwal replied:
I am thinking, either the struct Config could have componentconfig.KubeSchedulerConfiguration and not duplicate its field, or just duplicate fields whatever are needed in the struct, and not include componentconfig.KubeSchedulerConfiguration?

aveshagarwal commented on cmd/kube-scheduler/app/options/deprecated.go:
```diff
@@ -0,0 +1,116 @@
+		return nil
+	}
+	return nil
+}
```
why to have validate() when it always returns nil? also even if want this, i think just `return nil` is enough as thats what it does.

sttts replied:
it's the standard pattern. If you have actual validations (which we should have!), you put them in between.

aveshagarwal commented on cmd/kube-scheduler/app/options/deprecated.go:
```diff
@@ -0,0 +1,116 @@
+	if o == nil {
+		return
+	}
+
```
i think we would want to document that all these deprecated flags can still be configured as part of the config file, right?

sttts replied:
Even worse: they are effective only if no config file is given. This is old behaviour before this PR. We have to keep it for backwards compatibility. But actually we want to get rid of all of them ASAP to remove this odd behaviour.

aveshagarwal replied:
can you help me understand, why have you combined deprecated fields with non-deprecated fields here?

aveshagarwal commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,167 @@
+		componentConfig.HealthzBindAddress = ""
+		componentConfig.MetricsBindAddress = ""
+		return nil
+	}
```
I'd prefer to move above `if` block to ApplyTo.

aveshagarwal commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,167 @@
+		return fmt.Errorf("invalid healthz address: %v", err)
+	}
+	if err := updateInsecureServingOptionsFromAddress(o.Metrics, componentConfig.MetricsBindAddress); err != nil {
+		return fmt.Errorf("invalid metrics address: %v", err)
```
The above looks like cyclic, calling  updateInsecureServingOptionsFromAddress here and then calling updateAddressFromInsecureServingOptions in applyTo.

sttts replied:
@hzxuzhonghu had the same comment above. We call `updateInsecureServingOptionsFromAddress` to set the options from the componentconfig. We call `updateAddressFromInsecureServingOptions` to
- copy the address from the options to the componentconfig
- erase the componentconfig address if the respective `*InsecureServingOptions` is `nil` (a pattern we use to disable optional part of the configuration in custom plumbing, e.g. in Origin).

aveshagarwal commented on cmd/kube-scheduler/app/options/deprecated.go:
```diff
@@ -0,0 +1,116 @@
+
+// DeprecatedOptions contains deprecated options and their flags.
+// TODO remove these fields once the deprecated flags are removed.
+type DeprecatedOptions struct {
```
I am really trying to address what is the benefit of introducing this new DeprecatedOptions? Why can't we just use those fields directly from componentconfig.KubeSchedulerConfiguration?

sttts replied:
Because they should go away. They are legacy, so this PR split them from the code that has the final shape.

aveshagarwal replied:
ok now I see I agree here.

aveshagarwal commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,180 @@
+	CombinedInsecureServing *CombinedInsecureServingOptions
+	Authentication          *apiserveroptions.DelegatingAuthenticationOptions
+	Authorization           *apiserveroptions.DelegatingAuthorizationOptions
+	Deprecated              *DeprecatedOptions
```
I would really to understand advantages of Deprecated and not fetching its fields  directly from ComponentConfig above?

sttts replied:
`Deprecated` captures the code that should go away from the code that is in the future shape.

The fields here exist for the same reason they existed before, all this is just copied, separated for the upper reason, i.e. to hide the ugly: https://github.com/kubernetes/kubernetes/pull/60372/files#diff-f072139971e3329bef6da077a5a03aceL89

aveshagarwal commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,180 @@
+func newDefaultComponentConfig() (*componentconfig.KubeSchedulerConfiguration, error) {
+	scheme := runtime.NewScheme()
+	componentconfig.AddToScheme(scheme)
+	componentconfigv1alpha1.AddToScheme(scheme)
```
in loadConfig above you created a new scheme locally and here are we are doing again?

sttts replied:
fixed.

sttts commented on cmd/controller-manager/app/options/insecure_serving.go:
```diff
@@ -61,28 +64,18 @@ func (s *InsecureServingOptions) AddFlags(fs *pflag.FlagSet) {
 // ApplyTo adds InsecureServingOptions to the insecureserverinfo amd kube-controller manager configuration.
 // Note: the double pointer allows to set the *InsecureServingInfo to nil without referencing the struct hosting this pointer.
-func (s *InsecureServingOptions) ApplyTo(c **genericcontrollermanager.InsecureServingInfo, cfg *componentconfig.KubeControllerManagerConfiguration) error {
+func (s *InsecureServingOptions) ApplyTo(c **genericcontrollermanager.InsecureServingInfo) error {
```
In a follow-up we actually want to factor these out into a more neutral place.

sttts commented:
@aveshagarwal addressed all your comments. ptal.
@mtaufen fyi. As a followup, when this merges, I would like to start the discussion of what we can learn from your kubelet experience.

aveshagarwal commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,175 @@
+		return nil, err
+	}
+
+	hhost, hport, err := splitHostIntPort(cfg.HealthzBindAddress)
```
as i understand, this may not be needed here. you could just use this function above in options/insecure_serving.go.

sttts replied:
What do you mean by "this function above"?

aveshagarwal commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -0,0 +1,168 @@
+	portInt, err := strconv.Atoi(port)
+	if err != nil {
+		return fmt.Errorf("invalid non-numerical healthz port %q", port)
+	}
```
you could use splitHostIntPort here too.

sttts replied:
done

aveshagarwal commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -0,0 +1,175 @@
+	o.CombinedInsecureServing.AddFlags(fs)
+	o.Authentication.AddFlags(fs)
+	o.Authorization.AddFlags(fs)
+	o.Deprecated.AddFlags(fs, &o.ComponentConfig)
```
IMHO, it feels confusing to me as combining deprecated and non-deprecated fields unless i am missing something.

sttts replied:
alternative name?

aveshagarwal replied:
O.Deprecated.AddFlags and O.ComponentConfig.AddFlags?

sttts replied:
Not sure I can follow. I think we are talking at cross purposes.

`o.ComponentConfig` is the component config API type. We cannot attach methods and especially don't want to mix flags with them.



> It feels confusing to me as combining deprecated and non-deprecated fields

To my understanding all flags in `Deprecated` are deprecated, compare the old comment https://github.com/kubernetes/kubernetes/pull/60372/files#diff-f072139971e3329bef6da077a5a03aceL110. I only copy their definition here.

I agree that we should unify the way we technically deprecate the flags. Some have `MarkDeprecated` (and therefore disappear from `--help`), some don't.

aveshagarwal commented on cmd/kube-scheduler/app/options/plumbing.go:
```diff
@@ -0,0 +1,152 @@
+/*
```
Is there any benefit of creating this new file and not have it as part of options/options.go?

sttts replied:
no, there isn't. merged them.

sttts commented:
@aveshagarwal addressed your comments. ptal.

sttts commented:
@aveshagarwal could you take another look?

aveshagarwal commented on cmd/kube-scheduler/app/server.go:
```diff
@@ -363,207 +106,23 @@ through the API as necessary.`,
-	client, leaderElectionClient, eventClient, err := createClients(config.ClientConnection, master)
-	if err != nil {
-		return nil, err
+	if cz, err := configz.New("componentconfig"); err == nil {
```
why is `if len(config.MetricsBindAddress) > 0 || len(config.HealthzBindAddress) > 0`  removed for this if block?

sttts replied:
It does not hurt to set the configz struct if no http server is installed. This follows the pattern e.g. found in the controller manager.

aveshagarwal commented:
@sttts I dont have any more comments. lgtm. I am cc'ing @bsalamat for final review/approval/lgtm.

sttts commented:
@bsalamat rebased and squashed. Lgtm'ed by @aveshagarwal. Can you do a final pass?

k82cn commented:
/approve 

Approve for scheduler part, thanks :)

aveshagarwal commented:
@sttts do you wanna sqaush commits further? and I will lgtm it.

aveshagarwal commented:
/lgtm

sttts commented:
@aveshagarwal the 4 commits are pretty independent. Wouldn't squash them.

sttts commented:
The kubeadm change is a trival import change. Overriding approval (if the bot lets me).

sttts commented:
/retest
